### PR TITLE
Separate helm and startup commands

### PR DESCRIPTION
**Pull request by gilesknap**
- Split build-ioc command into build-helm and build-startup
- add  generic_image:  into the entity yaml and propgate that to the helm chart
- add a schema specification mode line to entity yaml
- validate against the schema in build-helm

### CODE REVIEW

**Comment by gilesknap** [comment]
@thomascobb sorry this PR may have got quite big. It now has the multiple definitions support.
I'm pretty happy with it so its up to you if you want to review.

File: Pipfile [hunk 1]
```diff
@@ -27,7 +27,8 @@ autopep8 = "*"
 # Pins to make lockfile usable on multiple Python versions and platforms
 typing-extensions = "*"
 # All other package requirements from setup.cfg
-ibek = {editable = true, path = "."}
+ibek = {editable = true,path = "."}
+jsonschema = "*"
```
Comment (thomascobb) [inline Pipfile:31]: This should go in setup.cfg

File: docs/explanations/entities.rst [hunk 2]
```diff
@@ -258,12 +258,13 @@ The ibek commands to progress through the file sequence above are as follows
...
-        - ``ibek build-ioc  <support>.ibek.yaml <ioc>.<container>.yaml``
+        - ``ibek build-helm  <container>.schema.json <ioc>.<container>.yaml``
           run at IOC helm chart generation time. This generates a helm chart
-          with ``<ioc>.<container>.yaml`` in its config folder
+          with ``<ioc>.<container>.yaml`` in its config folder and validates it
+          against ``<container>.schema.json``
     *   - 6
         - IOC startup script
-        - ``ibek make-ioc <ioc>.<container>.yaml ...``. Run at IOC startup time in the
+        - ``ibek make-startup <ioc>.<container>.yaml ...``. Run at IOC startup time in the
```
Comment (thomascobb) [inline docs/explanations/entities.rst:267]: Is this `build-startup`?

File: ibek/helm.py [hunk 3]
```diff
@@ -2,46 +2,86 @@
...
+            if len(parts) > 1:
+                schema_url = url_f + str(Path(parts[1]).absolute()).strip()
+
+            with urllib.request.urlopen(schema_url) as url:
+                entity_schema = json.loads(url.read().decode())
+
+        except Exception:
+            log.error(
+                f"Error getting schema for {ioc_instance_yaml}. "
+                "make sure it has '# yaml-language-server:'"
```
Comment (thomascobb) [inline ibek/helm.py:54]:
```suggestion
                "make sure it has '# yaml-language-server: $schema='"
```

File: ibek/ioc.py [hunk 4]
```diff
@@ -171,6 +171,7 @@ class IOC:
     ioc_name: A[str, desc("Name of IOC")]
     description: A[str, desc("Description of what the IOC does")]
     entities: A[Sequence[Entity], desc("List of classes of Entity this IOC supports")]
+    generic_image: A[str, desc("The generic IOC container image registry URL")]
```
Comment (thomascobb) [inline ibek/ioc.py:174]: Do we like the name generic? Would `ioc_image` or `container_name` be better?

File: Pipfile (hunk 1 above)
Comment (gilesknap) [reply Pipfile:31]: Thanks this is an easy thing to forget and we dont have checks for it on github. We do have dls-py3 internally. Should we do something about this?

File: docs/explanations/entities.rst (hunk 2 above)
Comment (gilesknap) [reply docs/explanations/entities.rst:267]: Yes

File: ibek/ioc.py (hunk 4 above)
Comment (gilesknap) [reply ibek/ioc.py:174]: it is a container image not a container and it is a generic ioc image, not a ioc instance. There is no such thing as a ioc instance image so maybe shortening to ioc_image is OK. 

But I will go with generic_ioc_image for now. Feel free to override me.

Comment (gilesknap) [reply ibek/ioc.py:174]: I've changed my mind already.

This class is all about generic IOC and if I use generic_ioc_image I should also use generic_ioc_name.

Instead how about changing the classname to GenericIOC ?

Comment (gilesknap) [reply ibek/ioc.py:174]: hmmm. Nope. IOC derived classes are Generic and IOC dervided instances are ioc instances so IOC is better. :-)

File: Pipfile (hunk 1 above)
Comment (thomascobb) [reply Pipfile:31]: I'm happy to just rely on gitlab CI for this, rather than make github do it too

File: ibek/ioc.py (hunk 4 above)
Comment (gilesknap) [reply ibek/ioc.py:174]: I rewrote the docstring and desc fields and named the offending member generic_ioc_image

```
@dataclass
class IOC:
    """
    Used to load an IOC instance entities yaml file into memory using 
    IOC.deserialize(YAML().load(ioc_instance_yaml).

    Before loading the entities file all Entity classes that it contains
    must be defined in modules.py. This is achieved by deserializing all
    support module definitions files used by this IOC and calling 
    make_entity_classes().
    """

    ioc_name: A[str, desc("Name of IOC instance")]
    description: A[str, desc("Description of what the IOC does")]
    entities: A[Sequence[Entity], desc("List of entities this IOC instantiates")]
    generic_ioc_image: A[str, desc("The generic IOC container image registry URL")]
```

**Pull request merged**